source/AI.cpp

```diff
@@ -3351,6 +3410,39 @@ double AI::RendezvousTime(const Point &p, const Point &v, double vp)
 }
 
 
+// Searches every asteroid within the ship scan limit and returns the asteroid
+// closest to the ship.
+bool AI::TargetMinable(Ship &ship) const
+{
+	double scanRange = 10000. * ship.Attributes().Get("asteroid scan power");
+	double baseline = scanRange;
+	int64_t highestCost = (ship.GetTargetAsteroid()) ? ship.GetTargetAsteroid()->GetCost() : 0.;
+	if(!baseline)
+		return false;
+	bool closestAsteroid = Preferences::Has("Target asteroid based on");
+	for(auto &&asteroid : minables)
+	{
+		double metric = ship.Position().DistanceSquared(asteroid->Position());
+		bool targetBasedOnCost = !closestAsteroid;
+		targetBasedOnCost &= asteroid->GetCost() >= highestCost;
+		// Don't target asteroids outside of your scan range.
+		targetBasedOnCost &= metric < scanRange;
+		// If asteroid value is the same as highest then target the one closest to your ship.
+		if(targetBasedOnCost && asteroid->GetCost() == highestCost)
+			targetBasedOnCost &= metric < baseline;
+		bool targetBasedOnProximity = closestAsteroid && metric < baseline;
+		if(targetBasedOnProximity || targetBasedOnCost)
+		{
+			// Target closest asteroid or target the highest value asteroid.
+			ship.SetTargetAsteroid(asteroid);
+			highestCost = asteroid->GetCost();
+			baseline = metric;
+		}
+	}
```

Comment:
Not really a fan of this. How about:
```suggestion
	double scanRangeMetric = 10000. * ship.Attributes().Get("asteroid scan power");
	if(!scanRangeMetric)
		return false;
	const bool findClosest = Preferences::Has("Target asteroid based on");
	auto bestAsteroid = ship.GetTargetAsteroid();
	double bestScore = findClosest ? numeric_limits<double>::max() : 0.;
	auto GetDistanceMetric = [&ship](const Minable &minable) -> double {
		return ship.Position().DistanceSquared(minable.Position());
	};
	if(bestAsteroid)
	{
		if(findClosest)
			bestScore = GetDistanceMetric(*bestAsteroid);
		else
			bestScore = bestAsteroid->GetValue();
	}
	auto UpdateBestAsteroid = [&findClosest, &bestAsteroid, &bestScore, &GetDistanceMetric]
			(const shared_ptr<Minable> &minable) -> void
	{
		double newScore = findClosest ? GetDistanceMetric(*minable) : minable->GetValue();
		if(newScore == bestScore)
		{
			if(findClosest && minable->GetValue() > bestAsteroid->GetValue())
				bestAsteroid = minable;
			else if(!findClosest && GetDistanceMetric(*minable) < GetDistanceMetric(*bestAsteroid))
				bestAsteroid = minable;
		}
		else if((findClosest && newScore < bestScore) || (!findClosest && newScore > bestScore))
		{
			bestScore = newScore;
			bestAsteroid = minable;
		}
	};
	for(auto &&asteroid : minables)
		UpdateBestAsteroid(asteroid);
	if(bestAsteroid)
		ship.SetTargetAsteroid(bestAsteroid);
```

Reply:
I don't have a strong opinion on this; as long as it behaves similar I defer to your better experience with C++ here with regard to performance and lambdas.  The way I was looping over asteroids was a copy of the vanilla code in other areas which loop over asteroids (for example when targeting nearest in vanilla).

Since I can't commit this directly I will try to make your suggested changes locally and attribute you with authorship.